src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,173 @@
+        {
+            SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            SyntaxNode node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                if (diagnostic.Id == StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId)
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithOneParameterFix,
+                                ctx => FixCodeOneParameter(context.Document, root, node, ctx),
+                                equivalenceKey: MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithOneParameterFix),
+                            diagnostic);
+                }
+                else
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix,
+                                ctx => FixCodeTwoParameters(context.Document, root, node, ctx),
+                                equivalenceKey: MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix),
+                            diagnostic);
+                }
+            }
+        }
+
+        private static async Task<Document> FixCodeOneParameter(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode nodeToFix,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var typedNodeToFix =  semanticModel.GetOperation(nodeToFix) as IInvocationOperation;
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var fixComponents = GetFixComponents(generator, typedNodeToFix);
+
+            SyntaxNode lengthNode = generator
+                .MemberAccessExpression(
+                    fixComponents.StringArgument.Syntax, 
+                    generator.IdentifierName(nameof(string.Length)));
+
+            var startIndexNode = fixComponents.OriginalInnerArguments[0].Value.Syntax;
+
+            var lengthArgument = generator
+                .SubtractExpression(
+                    lengthNode,
+                    startIndexNode);
+
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod, 
+                fixComponents.StringArgument.Syntax,
+                startIndexNode, 
+                lengthArgument);
+
+            var newRoot = root
```

Comment:
it might not be a bad idea to have a simple comment for each line explaining what is being generated.